### ISSUE TITLE
False-positive on column type mismatch with constant enumeration

### ISSUE THREAD

**Issue by Firehed** [phpstan/phpstan-doctrine#176]
Given a Doctrine model roughly like this:
```php
/**
 * @Entity
 */
class Some
{
  private const STATE_FOO = 'foo';
  private const STATE_BAR = 'bar';

  /**
   * @var self::STATE_*
   * @Column
   */
  private string $state;
}
```

I get an error like this:

`Property Some::$state type mapping mismatch: database can contain string but property expects string.`

Removing the `@var` annotation fixes the database type mismatch error (though can cause others, of course). It seems like string columns aren't correctly inferring the parent type when the backing variable is defined as a constant enumeration. I assume this would also apply to other scalars but haven't tested.

Versions:
```
$ composer info phpstan/*
phpstan/phpstan                   0.12.71 PHPStan - PHP Static Analysis Tool
phpstan/phpstan-deprecation-rules 0.12.6  PHPStan rules for detecting usage of deprecated classes, methods, properties, constants and traits.
phpstan/phpstan-doctrine          0.12.27 Doctrine extensions for PHPStan
phpstan/phpstan-phpunit           0.12.17 PHPUnit extensions and rules for PHPStan
phpstan/phpstan-strict-rules      0.12.9  Extra strict and opinionated rules for PHPStan
```

**Comment by ondrejmirtes** [comment]
The message could be improved, what it should be saying is:

> Property Some::$state type mapping mismatch: database can contain string but property expects 'foo'|'bar'.

What's it saying is that you have no guards against the database returning some other value like 'baz' because you're simply using column type string. You'd need a custom Doctrine type to be really sure. See the docs about how to make phpstan-doctrine understand your custom type: https://github.com/phpstan/phpstan-doctrine#custom-types

I'm gonna keep this open so that the message is improved.

**Comment by Seldaek** [comment]
Something really weird I experienced in relation to this.. Using latest phpstan and phpstan-doctrine:

```
class User {
    const AVATAR_TYPES = [
        self::AVATAR_TYPE_GRAVATAR,
        self::AVATAR_TYPE_UPLOADED_IMAGE,
        self::AVATAR_TYPE_INITIAL
    ];
    const AVATAR_TYPE_GRAVATAR = 'gravatar';
    const AVATAR_TYPE_UPLOADED_IMAGE = 'uploaded_image';
    const AVATAR_TYPE_INITIAL = 'initial';
    const DEFAULT_AVATAR_TYPE = self::AVATAR_TYPE_GRAVATAR;

   /**
     * @ORM\Column(type="string", length=100, nullable=true)
     * @Assert\Choice(choices=User::AVATAR_TYPES, groups={"registration", "profile"})
     * @var null|User::AVATAR_TYPE_GRAVATAR|User::AVATAR_TYPE_UPLOADED|User::AVATAR_TYPE_INITIAL
     */
    protected $avatarType;
}
```

This is OK for phpstan, but if I declare the type using a wildcard `@var null|User::AVATAR_TYPE_*` I get `Property User::$avatarType type mapping mismatch: database can contain string|null but property expects  'gravatar'|'initial'|'uploaded_image'|null.`

I kinda understand the mapping mismatch error, but I don't see why listing all the constants one by one makes it work. Any clue @ondrejmirtes?

**Comment by Seldaek** [comment]
That's my understanding as well, hence the confusion :)

**Comment by ondrejmirtes** [comment]
Try `\PHPStan\dumpType($this->avatarType)` in both variants to see what PHPStan thinks about the type :)

**Comment by Seldaek** [comment]
The  `AVATAR_TYPE_*` one outputs `Dumped type: 'gravatar'|'initial'|'uploaded_image'|null` as expected, the other `Dumped type: *ERROR*` which.. I guess explains why there is no mapping mismatch.

I just noticed now that i tis using `User::AVATAR_TYPE_UPLOADED` instead of `User::AVATAR_TYPE_UPLOADED_IMAGE`. If I fix that then it reports the mapping mismatch too, sorry for the noise here. Now the question is why it didn't report the invalid constant as an error :)

**Comment by ondrejmirtes** [comment]
You should be getting error `'%s for property %s::$%s contains unresolvable type.'` on level 2 in case of an unknown constant.

**Comment by Seldaek** [comment]
Thanks for the assistance and sorry for the trouble, user error really, level 1 should not be allowed ;)
